### PR TITLE
[SPARK-49021][SS] Add support for reading transformWithState value state variables with state data source reader

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add support for reading transformWithState value state variables with state data source reader

Co-authored with @jingz-db 

### Why are the changes needed?
Changes are needed to integrate reading state reading with new operator metadata and state schema format for the value state types used in state variables within transformWithState

### Does this PR introduce _any_ user-facing change?
Yes

Users can now read valueState variables used in the `transformWithState` operator using the state data source reader.

```
spark
   .read
   .format("statestore")
   .option("operatorId", <operatorId>)
   .option("stateVarName", <varName>)
   .load(<state path>)

```

### How was this patch tested?
Added unit tests
```
===== POSSIBLE THREAD LEAK IN SUITE o.a.s.sql.streaming.TransformWithStateSuite, threads: ForkJoinPool.commonPool-worker-4 (daemon=true), Idle Worker Monitor for python3 (daemon=true), rpc-boss-3-1 (daemon=true), ForkJoinPool.commonPool-worker-5 (daemon=true), ForkJoinPool.commonPool-worker-3 (daemon=true), ForkJoinPool.commonPool-worker-2 (daemon=true), shuffle-boss-6-1 (daemon=true), ForkJoinPool.commonPool-worker-1 (daemon=true) =====
[info] Run completed in 2 minutes, 28 seconds.
[info] Total number of tests run: 42
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 42, failed 0, canceled 0, ignored 1, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
@jingz-db @HeartSaVioR - PTAL, thx !

**Comment by HeartSaVioR** [comment]
General comment: JIRA ticket seems to be mentioning list / map state. What's the right thing? Please make a correction in either thing.

**Comment by anishshri-db** [comment]
> General comment: JIRA ticket seems to be mentioning list / map state and PR title is mentioning value state. What's correct? Please make a correction.

Apologies - JIRA ticket had wrong description. Updated it

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala:239]: nits: Shall we add some message in the `require` here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala:209]: Do we call this `getStoreMetadataAndRunChecks` in `inferSchema` and in `getTable` twice so we will do `StateMetadataPartitionReader` twice? Is it possible that we only run this once and store it as a class variable?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala:135]: In what situation we will have `keyStateEncoderSpecOpt` being None? Shall we throw error in this case?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:89]: Haven't got the chance to read Eric's VCF integration PR yet. Will we need to call `getStore` after his PR is merged?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (jingz-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:192]: Shall we place this function and the row schema definition (code [here](https://github.com/apache/spark/pull/47574/files#diff-d65c7a77b70d4f549969ddd852d83e0a337bfa8c22e1b0a76f8e1ad5c96982d8R187)) into one single class/object so that all following TWS state vars can define their schema/unify function in the class? This way we can probably have better code readability for future List/Map state reader implementations.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala
Comment (jingz-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:1073]: Shall we also have another test cases for value state with TTL?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala:136]: Only the stream-stream join case. Added a small comment

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:89]: Yea might still be needed, but I need to merge back his change for the TTL changes to work

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:1073]: Yup - waiting for the vcf integration PR to merge

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala]: Done

**Comment by anishshri-db** [comment]
@HeartSaVioR @jingz-db - could you ptal ? Thx

**Review by jingz-db — APPROVED** [review]
Thanks for the refactor, the future changes on state reader integration will be cleaner!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:95]: I think it's redundant to create a column family as we will be able to restore all available column families during the load. Is this to consider the bad case of "CF does not exist" to the same with "no data"?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:165]: nit: `{ pair =>`. Let's use `{` for multi-lines which is more consistent with existing code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:232]: What's the intention of this change? Does this make any impact on planning phase of IncrementalExecution?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:29]: nit: not necessary?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:41]: ditto about the placement of private method

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:114]: nit: inline or explicitly follow the style guidance of constant. EXPECTED_COLUMN_TYPES or so.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:75]: Are "read change feed" and "read from transformWithState" mutually exclusive? If then, have we blocked the case of specifying both options?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala:273]: nit: no need to have `s` before the string.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:77]: It's a huge change of the value of TTL - if the existing test does not fail with this change, it sounds to me as the existing test is too loose.

**Review by HeartSaVioR — COMMENTED** [review]
First round.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala:111]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/metadata/StateMetadataSource.scala:232]: Its to avoid the exception being thrown if the path doesn't exist. I don't think we need any changes in `IncrementalExecution` - it should handle the case and return an empty Map to the caller as we expect ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala]: Yea - actually that test was not relying on these configs at all. But anyway - just moved it to the data source related suite as a separate processor.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:95]: I don't think that will work @HeartSaVioR - in this case - in the regular path, the operator `init` function is invoking the `createColFamilyIfAbsent` call that will add the encoding info to the map in the provider. So we need to emulate that here as well.

If I remove the check, the test fails as expected where its unable to find the encoder in the provider

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:95]: Ah OK, I missed to consider the path of "RocksDB state store provider". I only considered the path of "RocksDB". My bad.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:46]: Yea not supported currently. We will support it eventually. Added the block for now and corresponding tests

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala]: Made this stateless too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala]: Yea removed this dependence

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala]: Done

**Review by HeartSaVioR — APPROVED** [review]
+1 pending CI

**Comment by HeartSaVioR** [comment]
https://github.com/anishshri-db/spark/actions/runs/10609063268/job/29404193245

```
[info] *** 2 TESTS FAILED ***
[error] Failed: Total 11562, Failed 2, Errors 0, Passed 11560, Ignored 31, Canceled 1
[error] Failed tests:
[error] 	org.apache.spark.sql.execution.datasources.v2.state.StateDataSourceNegativeTestSuite
[error] (sql / Test / test) sbt.TestsFailedException: Tests unsuccessful
```

CI failure seems to be related.

**Comment by anishshri-db** [comment]
> CI failure seems to be related.
Yes thanks - fixed the issue. Should be resolved now. Will wait for CI to finish

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**
